File: src/main/java/com/gluck/jobtracker/DataGenerator.java
package com.gluck.jobtracker;

import com.gluck.jobtracker.repository.JobApplicationEntity;
import com.gluck.jobtracker.model.Status;
import com.gluck.jobtracker.repository.UserEntity;
import com.gluck.jobtracker.repository.ApplicationRepository;
import com.gluck.jobtracker.repository.UserRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.CommandLineRunner;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.time.LocalDate;
import java.util.List;

@Configuration
public class DataGenerator {

    @Value("${admin.username}")
    private String adminUser;

    @Value("${admin.password}")
    private String adminPass;

    @Bean
    public CommandLineRunner loadData(ApplicationRepository repository) {
        return args -> {
            if (repository.count() == 0L) {
                repository.saveAll(
                    List.of(
                        createJob(
                            "Junior Backend Engineer",
                            "Google",
                            Status.APPLIED,
                            LocalDate.of(2025, 12, 17),
                            "TEST DATA, [link], position in Helsinki, credible company"
                        ),
                        createJob(
                            "Kotlin Engineer",
                            "Local Startup N",
                            Status.INTERVIEWING,
                            LocalDate.of(2025, 12, 1),
                            "TEST DATA, gggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg"
                        ),
                        createJob(
                            "Spring Boot Pro",
                            "FInTech Corp",
                            Status.WISH_LIST,
                            LocalDate.of(2025, 12, 7),
                            null
                        )
                    )
                );
                System.out.println("Generated demo data for the job tracker.");
            }
        };
    }

    private JobApplicationEntity createJob(
        String position,
        String company,
        Status status,
        LocalDate dateApplied,
        String description
    ) {
        JobApplicationEntity entity = new JobApplicationEntity();
        entity.setPosition(position);
        entity.setCompanyName(company);
        entity.setStatus(status);
        entity.setDateApplied(dateApplied);
        entity.setDescription(description);
        return entity;
    }

    @Bean
    public CommandLineRunner loadUserData(UserRepository repository, PasswordEncoder passwordEncoder) {
        return args -> {
            if (repository.findByUsername(adminUser) == null) {
                UserEntity user = new UserEntity(
                        adminUser,
                        passwordEncoder.encode(adminPass),
                        "ADMIN"
                );
                repository.save(user);
                System.out.println("Generated ADMIN user.");
            } else {
                System.out.println("User " + adminUser + " already generated. Skipping");
            }
        };
    }
}

File: src/main/kotlin/com/gluck/jobtracker/DataGenerator.kt
package com.gluck.jobtracker

import com.gluck.jobtracker.model.JobApplicationEntity
import com.gluck.jobtracker.model.Status
import com.gluck.jobtracker.model.UserEntity
import com.gluck.jobtracker.repository.ApplicationRepository
import com.gluck.jobtracker.repository.UserRepository
import com.vaadin.copilot.shaded.bouncycastle.crypto.generators.BCrypt
import org.springframework.boot.CommandLineRunner
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.crypto.password.PasswordEncoder
import java.time.LocalDate
import kotlin.uuid.Uuid

@Configuration
class DataGenerator {

    @Bean
    fun loadData(repository: ApplicationRepository): CommandLineRunner {
        return CommandLineRunner {
            if (repository.count() == 0L) {
                repository.saveAll(
                    listOf(
                        JobApplicationEntity(
                            position = "Junior Backend Engineer",
                            companyName = "Google",
                            status = Status.APPLIED,
                            dateApplied = LocalDate.of(2025, 12, 17),
                            description = "TEST DATA, [link], position in Helsinki, credible company"
                        ),
                        JobApplicationEntity(
                            position = "Kotlin Engineer",
                            companyName = "Local Startup N",
                            status = Status.INTERVIEWING,
                            dateApplied = LocalDate.of(2025, 12, 1),
                            description = "TEST DATA, gggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg"
                        ),
                        JobApplicationEntity(
                            position = "Spring Boot Pro",
                            companyName = "FInTech Corp",
                            status = Status.WISH_LIST,
                            dateApplied = LocalDate.of(2025, 12, 7)
                        ),
                    )
                )
                println("Generated demo data for the job tracker.")
            }
        }
    }

    @Bean
    fun loadUserData(repository: UserRepository, passwordEncoder: PasswordEncoder): CommandLineRunner {
        return CommandLineRunner {
            if (repository.findByUsername("admin") == null) {
                repository.save(
                    UserEntity(
                        username = "admin",
                        password = passwordEncoder.encode("superPassword")!!, // I know it shouldn't be here
                        role = "ADMIN"                                                      // I am working on keeping it a secret
                    )
                )
                println("Generated ADMIN user.")
            }
        }
    }

}